core/trino-main/src/main/java/io/trino/execution/buffer/SpoolingExchangeOutputBuffer.java

```diff
@@ -166,34 +170,43 @@ public void enqueue(int partition, List<Slice> pages)
     @Override
     public void setNoMorePages()
     {
-        stateMachine.noMorePages();
-        destroy();
+        if (!stateMachine.noMorePages()) {
+            return;
+        }
+
+        exchangeSink.finish().whenComplete((value, failure) -> {
+            if (failure != null) {
+                stateMachine.fail(failure);
+            }
+            else {
+                stateMachine.finish();
+            }
+            updateMemoryUsage(0);
+        });
     }
 
     @Override
     public void destroy()
     {
-        if (stateMachine.finish()) {
-            try {
-                exchangeSink.finish();
-            }
-            finally {
-                updateMemoryUsage(exchangeSink.getMemoryUsage());
-            }
-        }
+        // Abort the buffer if it hasn't been finished. This is possible when a task is cancelled early by the coordinator.
```

Comment:
that description is confusing:
```
This is possible when a task is cancelled early by the coordinator.
```
and
```
Task cancellation is not supported as the task output is expected to be deterministic.
```
Both can't be true at same time, right?

Reply:
The task cancellation is not expected to be requested by coordinator. It can only be requested if there's a bug in the scheduler. However if this situation happens (e.g.: due to a bug) it is safer to invalidate the buffer with `abort` to avoid publishing incomplete data to the exchange service. 

Added one more sentence to elaborate it.